### PR TITLE
gh-101678: refactor the math module to use special functions from c11

### PR DESCRIPTION
**Pull request by skirpichev**
Shouldn't affect users, hence no news.

* Issue: gh-101678

Automerge-Triggered-By: GH:mdickinson

### CODE REVIEW

**Comment by mdickinson** [comment]
What's the benefit of doing this? The downside is that we potentially introduce a quality regression on some platforms.

**Comment by skirpichev** [comment]
> What's the benefit of doing this?

Less C code to maintain, perhaps?

BTW, for erf/erfc there are no failures so far.

**Comment by corona10** [comment]
> What's the benefit of doing this? The downside is that we potentially introduce a quality regression on some platforms.

If the patch satisfies C11 compiler requirement and [PEP 11](https://peps.python.org/pep-0011/) support, it could be acceptable.

**Comment by mdickinson** [comment]
@corona10 Yes, the "this" in my comment could have been clearer.

So long as we keep this PR focused on removing what's now effectively dead code (the code that depends on `HAVE_LOG2`, `HAVE_ERF`, etc.), that's fine. It's the replacement of our own `gamma` and `lgamma` code that should be avoided - we'd be replacing tested, working code with code of unknown quality.

FWIW, speaking as the original author of the `gamma` and `lgamma` functions and maintainer of the `math` module, the maintenance burden from that code has been essentially zero since it was originally added.

**Comment by mdickinson** [comment]
@skirpichev Thanks for the PR! I'll have time to review properly this evening (UTC+0).

**Comment by skirpichev** [comment]
> So long as we keep this PR focused on removing what's now effectively dead code

That was essentially the target of the issue.

Thank you for review and I'm sorry for a missed issue/pr (adding c11 keyword to search was too restrictive this time...)

**Comment by skirpichev** [comment]
Maybe using math_1 wrapper for erf/erfc is more safe?

File: Modules/mathmodule.c
Comment (mdickinson) [inline Modules/mathmodule.c]: Please keep the braces here (for PEP 7 compliance).

File: Modules/mathmodule.c
Comment (mdickinson) [inline Modules/mathmodule.c]: Grammar nitpick:
```suggestion
/* Implementation of the real gamma function.  Kept here to work around
```

**Comment by mdickinson** [comment]
> Maybe using math_1 wrapper for erf/erfc is more safe?

I don't think it's needed, since `erf` and `erfc` can't overflow, and are valid over the entire real line.

**Review by mdickinson — COMMENTED** [review]
LGTM aside from two tiny nitpicks.

**Comment by skirpichev** [comment]
> I don't think it's needed, since erf and erfc can't overflow, and are valid over the entire real line.

Same is true of the tanh, for example, isn't?

**Comment by skirpichev** [comment]
I would suggest merging math_1 and math_1_to_whatever (not used anywhere else).  The comment before math_1_to_whatever does mention math_1 instead.  Ditto for math_1_to_whatever.

**Comment by mdickinson** [comment]
> I would suggest merging math_1 and math_1_to_whatever (not used anywhere else).

Sounds reasonable, but for ease of review let's have a separate PR for that, please! I'd like to get this one in, and that gets harder if the scope keeps changing.

File: Modules/_math.h
Comment (arhadthedev) [inline Modules/_math.h]: Python sometimes changes bugtracker systems (SourceForge bugtracker → Roundup → GitHub Issues) invalidating all relative issue numbers on export. So we need to specify a full number.

```suggestion
    /* Some platforms (e.g. MacOS X 10.8, see gh-59682) supply a log1p function
```

**Comment by mdickinson** [comment]
> Same is true to the tanh, for example, isn't?

Very likely, yes. Note that that's _not_ a reason to change `tanh` as it currently is. The libm wrappers portion of the math module is highly stable, and we should be extremely conservative with changes.

File: Modules/_math.h
Comment (arhadthedev) [reply Modules/_math.h]: Side note: Roundup migration was done [in 2008 (Python 2.6)](https://docs.python.org/2.6/whatsnew/2.6.html#new-issue-tracker-roundup), GitHub  migration was done [last year](https://discuss.python.org/t/github-issues-are-now-live/14967). And nobody knows when and why the next migration will take place.

Comment (skirpichev) [reply Modules/_math.h]: Thanks.  I thought that such references were fixed (that's how I did conversion for sympy) in sources too and now there is some canonical way to address issues.

**Review by mdickinson — APPROVED** [review]
LGTM. I'll merge when CI completes.

**Comment by mdickinson** [comment]
Aargh. And now I remember (again) why I never use the auto-merge label - there's no opportunity to edit the final commit message. :-(

**Comment by skirpichev** [comment]
Thanks for the hint.  Next time I'll adjust pr title/body to follow review.

**Comment by vstinner** [comment]
Nice change.

> Shouldn't affect users, hence no news.

In the past, I tried to document these "build changes", just in case if someone is affected. Like:

> The C99 functions snprintf() and vsnprintf() are now required to build Python. (Contributed by Victor Stinner in bpo-36020.)

https://docs.python.org/dev/whatsnew/3.10.html#build-changes

**Comment by skirpichev** [comment]
c11 compiler and stdlib (without optional features) were required [since 3.11](https://docs.python.org/dev/whatsnew/3.11.html#build-changes), so there is nothing new in that sense.

**Pull request merged**
